### ISSUE TITLE
Windows Server 2003/2008 DLL loader lock deadlock

### ISSUE THREAD

**Issue by reganheath** [weidai11/cryptopp#435]
**TL;DR:** The 5.6.5 change to Integer::Zero() has introduced a DLL loader lock deadlock on windows server 2003 and server 2008.

I build crypto++ into a DLL (called "transformd") which I link to several executables (import library/dll combo).

Since upgrading to Crypto++ 5.6.5 these executables hang when I try to run them on windows server 2003 or 2008 (but not 2012 or 2016).

Running a debug build in windbg shows the following stack trace:

```
ntdll!NtWaitForKeyedEvent+0x15
ntdll!RtlFreeUserStack+0x2bb
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\Windows\syswow64\kernel32.dll - 
ntdll!RtlRegisterWait+0x183
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\Htape\DR\bin\MSVCR120D.dll - 
kernel32!RegisterWaitForSingleObject+0x4f
MSVCR120D!Concurrency::set_task_execution_resources+0x4298
MSVCR120D!Concurrency::details::_SetUnobservedExceptionHandler+0x7fe
MSVCR120D!Concurrency::details::_SetUnobservedExceptionHandler+0x977
MSVCR120D!Concurrency::Scheduler::Create+0x17c
MSVCR120D!Concurrency::Scheduler::Create+0x148a
MSVCR120D!Concurrency::Scheduler::Create+0x6b
MSVCR120D!Concurrency::Context::CurrentContext+0x25
MSVCR120D!Concurrency::details::platform::GetCurrentThreadId+0x980
*** ERROR: Symbol file could not be found.  Defaulted to export symbols for C:\Htape\DR\bin\MSVCP120D.dll - 
MSVCR120D!Concurrency::critical_section::lock+0x13
MSVCP120D!Thrd_yield+0x159
MSVCP120D!Mtx_lock+0xe
transformd!std::_Mtx_lockX+0x10 [c:\program files (x86)\microsoft visual studio 12.0\vc\include\thr\xthread @ 68]
transformd!std::_Mutex_base::lock+0x17 [c:\program files (x86)\microsoft visual studio 12.0\vc\include\mutex @ 41]
transformd!std::lock_guard<std::mutex>::lock_guard<std::mutex>+0x20 [c:\program files (x86)\microsoft visual studio 12.0\vc\include\mutex @ 184]
transformd!CryptoPP::Singleton<CryptoPP::Integer,CryptoPP::NewObject<CryptoPP::Integer>,0>::Ref+0xcb [s:\src\c\cryptopp\misc.h @ 296]
transformd!CryptoPP::`anonymous namespace'::`dynamic initializer for 's_zero''+0x24 [s:\src\c\cryptopp\integer.cpp @ 3031]
MSVCR120D!initterm+0x1a
transformd!_CRT_INIT+0xc3 [f:\dd\vctools\crt\crtw32\dllstuff\crtdll.c @ 295]
transformd!__DllMainCRTStartup+0xac [f:\dd\vctools\crt\crtw32\dllstuff\crtdll.c @ 502]
transformd!_DllMainCRTStartup+0x1f [f:\dd\vctools\crt\crtw32\dllstuff\crtdll.c @ 472]
ntdll!RtlQueryEnvironmentVariable+0x245
ntdll!LdrResSearchResource+0xa36
ntdll!RtlGetNtVersionNumbers+0x9b
ntdll!RtlSetUnhandledExceptionFilter+0x50
ntdll!LdrInitializeThunk+0x10
```

A google found this post which I think describes the issue: [Microsoft Connect Issue 809005, Deadlock when locking std::mutex during DllMain static initialization](https://connect.microsoft.com/VisualStudio/feedback/details/809005/deadlock-when-locking-std-mutex-during-dllmain-static-initialization).

Essentially my exe loads my dll which follows the default dll initialisation path calling CRT_INIT to initialise global variables, like s_zero, which locks a std::mutex and sends us into the "not dllmain safe" concurrency code, which I suspect is waiting for something else to initialise - but as we hold the loader lock that thing cannot initialse, thus the deadlock.

**Comment by noloader** [comment]
Thanks @reganheath,

Yeah, that sounds about right. Microsoft did not list synchronization and threads as unsupported; rather, they claimed [mutexes and threads were supported C++ features](https://msdn.microsoft.com/en-us/library/hh567368.aspx). They really should have listed them as broken, like the [Dynamic Initialization and Destruction with Concurrency](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2008/n2660.htm) problem.

I still don't have a canonical reference from MSDN. Instead, we see an occasional report like this. Also see [Does std::call_once work under VS2013 and below?](https://stackoverflow.com/q/43033009/608639) on Stack Overflow.

I believe we fixed it in Master under [Issue 373, Loader lock on WinXP](https://github.com/weidai11/cryptopp/issues/373). My first question is, do you experience the issue under Master? If you do experience it under Master, then that's really bad news for Microsoft users. We will have to back out the attempts to get a handle on the problem, and go back to all the earlier memory leaks.

**Comment by reganheath** [comment]
Hi @noloader 

Thanks for the reply.

Unfortunately I cannot test Master, however I did drop the relevant code changes from Integer.cpp into 5.6.5 and it does work for me.  Visual Studio 2013 does define HAVE_MSC_INIT_PRIORITY and so #pragma init_seg is being used - so, no memory leaks either :)

**Comment by noloader** [comment]
Thanks @reganheath,

So it sounds like we are OK now. Is it safe to close this report?

The reason I ask is I want to ensure I did not do harm instead of good with the change. Its been challenging due to lack of documentation from Microsoft.

**Comment by reganheath** [comment]
Yeah, @noloader, I reckon you should close it.  I mean, my test wasn't the full Master code, just a snippet pulled from Integer.cpp, but that snippet did work, at least for me, on Server 2008 R2.

**Comment by noloader** [comment]
@reganheath,

Ack, thanks. WIll do. If we encounter problems, then I'll re-open this.

Thanks again.
